src/main/java/picard/illumina/ClusterDataToSamConverter.java

```diff
@@ -230,26 +272,61 @@ private SAMRecord createSamRecord(final ReadData readData, final String readName
         }
 
         final SAMRecord firstOfPair = createSamRecord(
-            cluster.getRead(templateIndices[0]), readName, cluster.isPf(), true, unmatchedBarcode, molecularIndexes, molecularIndexQualities);
+                cluster.getRead(templateIndices[0]),
+                readName,
+                cluster.isPf(),
+                true,
+                unmatchedBarcode,
+                barcodeQuality,
+                molecularIndexes,
+                molecularIndexQualities);
         ret.records[0] = firstOfPair;
 
+
         SAMRecord secondOfPair = null;
 
-        if(isPairedEnd) {
-            secondOfPair  = createSamRecord(
-                cluster.getRead(templateIndices[1]), readName, cluster.isPf(), false, unmatchedBarcode, molecularIndexes, molecularIndexQualities);
+        if (isPairedEnd) {
+            secondOfPair = createSamRecord(
+                    cluster.getRead(templateIndices[1]),
+                    readName,
+                    cluster.isPf(),
+                    false,
+                    unmatchedBarcode,
+                    barcodeQuality,
+                    molecularIndexes,
+                    molecularIndexQualities);
             ret.records[1] = secondOfPair;
         }
 
         if (adapterMarker != null) {
             // Clip the read
             if (isPairedEnd) {
                 adapterMarker.adapterTrimIlluminaPairedReads(firstOfPair, secondOfPair);
-            }
-            else {
+            } else {
                 adapterMarker.adapterTrimIlluminaSingleRead(firstOfPair);
             }
         }
         return ret;
     }
+
+    private String getBarcodeQuality(ClusterData cluster) {
+        final String[] barcodeQ = new String[sampleBarcodeIndices.length];
```

Comment:
can use a stringjoiner to avoid creating array
```
        final StringJoiner barcodeQ = new StringJoiner("~");

        for (int sampleBarcodeIndice : sampleBarcodeIndices) {
            barcodeQ.add(SAMUtils.phredToFastq(cluster.getRead(sampleBarcodeIndice).getQualities()));
        }
        return barcodeQ.toString();
```

Reply:
TIL